packages/snaps-controllers/src/utils.ts

```diff
@@ -375,3 +376,97 @@ export function debouncePersistState(
     );
   };
 }
+
+/**
+ * Handlers allowed for tracking.
+ */
+export const TRACKABLE_HANDLERS = Object.freeze([
+  HandlerType.OnHomePage,
+  HandlerType.OnInstall,
+  HandlerType.OnNameLookup,
+  HandlerType.OnRpcRequest,
+  HandlerType.OnSignature,
+  HandlerType.OnTransaction,
+  HandlerType.OnUpdate,
+] as const);
+
+/**
+ * A union type representing all possible trackable handler types.
+ */
+export type TrackableHandler = (typeof TRACKABLE_HANDLERS)[number];
+
+/**
+ * Throttles event tracking calls per unique combination of parameters.
+ *
+ * @param fn - The tracking function to throttle.
+ * @param timeout - The timeout in milliseconds. Defaults to 60000 (1 minute).
+ * @returns The throttled function.
+ */
+export function throttleTracking(
+  fn: (
+    snapId: SnapId,
+    handler: TrackableHandler,
+    success: boolean,
+    origin: string,
+  ) => void,
+  timeout = 60000,
+) {
+  const previousCalls = new Map<string, number>();
+  const pendingCalls = new Map<
+    string,
+    {
+      args: [SnapId, TrackableHandler, boolean, string];
+      timer: ReturnType<typeof setTimeout> | null;
+    }
+  >();
+
+  return (
+    snapId: SnapId,
+    handler: TrackableHandler,
+    success: boolean,
+    origin: string,
+  ): void => {
+    const key = `${snapId}${handler}${success}${origin}`;
+    const now = Date.now();
+    const lastCall = previousCalls.get(key) ?? 0;
+    const args: [SnapId, TrackableHandler, boolean, string] = [
+      snapId,
+      handler,
+      success,
+      origin,
+    ];
+
+    if (!previousCalls.has(key) || now - lastCall >= timeout) {
```

Comment:
```suggestion
    if (now - lastCall >= timeout) {
```

I don't think we need the extra branch since we have a default for `lastCall`